### PR TITLE
Fix mlflow evaluate using xgboost model

### LINKED ISSUE #12516: [BUG] SHAP evaluation fails due to mlflow.exceptions.MlflowException: Incompatible input types for column columnname. Can not safely convert float64 to int32.
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

mlflow, version 2.14.1

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **Python version**: 3.10

### Describe the problem

During evaluation I am getting an error:
mlflow.exceptions.MlflowException: Incompatible input types for column columnname. Can not safely convert float64 to int32.

There are columns in my dataframe with numpy types int8, int16, int32. However it seems inside the SHAP evaluation process they are converted to float64, then back in mlflow code it tries to compare them to the schema but the schema expects ints hence the error.

It's not clear to me if I am doing something wrong, or this is an issue with the mlflow code, or an issue in the SHAP code. I would appreciate any guidance.

### Tracking information

_No response_

### Code to reproduce issue

```
from pathlib import Path

import pandas as pd
from sklearn.model_selection import train_test_split

import mlflow
import xgboost as xgb
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

mlflow.set_tracking_uri("http://localhost:5000")
logging.getLogger('mlflow').setLevel(logging.DEBUG)
df = pd.read_parquet(Path(__file__).parent / 'df_preprocessed.parquet')
target_col = 'target'
X = df.copy()
y = X.pop(target_col)

logger.info('Splitting %d rows of data into train, validation, test datasets.', len(X))
rand = 42  # random number seed.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=rand)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=rand)
X_train, X_debug, y_train, y_debug = train_test_split(X_train, y_train, test_size=0.1, random_state=rand)

# setup experiment tracking if it hasn't already been setup
experiment_name = 'expname'
tags = dict(project_name='proj', team='myteam')
existing_experiments = mlflow.search_experiments(filter_string=f"name = '{experiment_name}'")
if not existing_experiments:
	print("Experiment doesn't exist so we will create it")
	mlflow.create_experiment(name=experiment_name, tags=tags)
mlflow.set_experiment(experiment_name)
mlflow.autolog(log_input_examples=True)

import warnings

warnings.filterwarnings(
	action='ignore', message='.*Failed to determine whether UCVolumeDatasetSource can resolve source information.*'
)

input_data_url = 's3://bucketname/folder/folder'
train_dataset = mlflow.data.from_pandas(
	pd.concat([y_train, X_train], axis='columns'),
	source=input_data_url,
	name='dataset name',
	targets=target_col,
)

xgb_model = xgb.XGBClassifier(tree_method='hist', objective='binary:logistic', random_state=42)

active_run = mlflow.start_run()
mlflow.log_input(train_dataset, context="train")
xgb_model.fit(X_train, y_train)

datas = [
	y_val,
	X_val,
]
val_dataset = mlflow.data.from_pandas(
	pd.concat(datas, axis='columns'),
	source=input_data_url,
	name='dataset - validation data',
	targets=target_col,
)

# evaluate
model_uri = mlflow.get_artifact_uri("model")
result = mlflow.evaluate(model=model_uri, data=val_dataset, model_type="classifier")
print('Result: !!!\n' + str(result))
mlflow.end_run()
```

### Stack trace

```
2024/06/28 15:28:54 DEBUG mlflow.models.evaluation.default_evaluator: 
Traceback (most recent call last):
  File "C:\dev\ws\proj\.venv\lib\site-packages\mlflow\pyfunc\__init__.py", line 698, in _validate_prediction_input
    data = _enforce_schema(data, input_schema, flavor)
  File "C:\dev\ws\proj\.venv\lib\site-packages\mlflow\models\utils.py", line 1028, in _enforce_schema
    _enforce_named_col_schema(pf_input, input_schema)
  File "C:\dev\ws\proj\.venv\lib\site-packages\mlflow\models\utils.py", line 735, in _enforce_named_col_schema
    new_pf_input[name] = _enforce_mlflow_datatype(name, pf_input[name], input_type)
  File "C:\dev\ws\proj\.venv\lib\site-packages\mlflow\models\utils.py", line 695, in _enforce_mlflow_datatype
    raise MlflowException(
mlflow.exceptions.MlflowException: Incompatible input types for column columnname. Can not safely convert float64 to int32.
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12599?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12599/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12599
```

</p>
</details>

### Related Issues/PRs

Resolve #12516

[robmcd](https://github.com/robmcd) opened this issue last week · 2 comments

### What changes are proposed in this pull request?
The root cause of the problem is that mlflow autologging uses `mlflow.xgboost` to log the model, and we fail to extract the real xgboost raw model in such case. Shap explainer needs to use the original raw model's predict function as the sampled data is converted to float64 type and shouldn't be verified against the model signature. This PR fixes the problem.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/models/evaluation/default_evaluator.py [hunk 1]
```diff
@@ -131,11 +131,22 @@ def _infer_model_type_by_labels(labels):
...
             return model_loader_module, model._model_impl.sklearn_model
+        # extract the raw model of xgboost flavor so that shap explainer uses the raw model
+        # instead of the wrapper and skip data schema validation
+        try:
+            from mlflow.xgboost import _XGBModelWrapper
+
+            if isinstance(model._model_impl, _XGBModelWrapper):
+                return model_loader_module, model._model_impl.xgb_model
+        except ImportError:
+            pass
```
Comment (WeichenXu123) [inline mlflow/models/evaluation/default_evaluator.py:149]: Thoughts: shall we add an interface like `get_raw_model` for `_XXXModelWrapper` class ? So that in `mlflow/models/evaluation/default_evaluator.py` we can simply call `model._model_impl.get_raw_model`

Comment (serena-ruan) [reply mlflow/models/evaluation/default_evaluator.py:149]: This is a good idea!

File: temp.zip
Comment (B-Step62) [inline temp.zip:1]: unrelated file?

Comment (serena-ruan) [reply temp.zip:1]: Ahhh this is the weird databricks githook... Let me delete it 😓

File: mlflow/sklearn/__init__.py [hunk 2]
```diff
@@ -518,6 +518,12 @@ def __init__(self, sklearn_model):
             if fn := getattr(self.sklearn_model, predict_fn, None):
                 setattr(self, predict_fn, fn)
 
+    def get_raw_model(self):
+        """
+        Returns the underlying scikit-learn model.
+        """
+        return self.sklearn_model
```
Comment (WeichenXu123) [inline mlflow/sklearn/__init__.py:525]: similarly, we can add a `_wrapped_predict_fn` function into the returned raw model object.

File: mlflow/models/evaluation/default_evaluator.py [hunk 3]
```diff
@@ -171,12 +174,26 @@ def _extract_predict_fn(
...
+                    def wrapped_predict_fn(data, *args, **kwargs):
+                        return raw_model.predict(
+                            xgboost.DMatrix(data), *args, validate_features=False, **kwargs
+                        )
+
+                    return wrapped_predict_fn
+                elif isinstance(raw_model, xgboost.XGBModel):
+                    return partial(raw_model.predict, validate_features=False)
+                else:
+                    return raw_model.predict
```
Comment (WeichenXu123) [inline mlflow/models/evaluation/default_evaluator.py:191]: With this https://github.com/mlflow/mlflow/pull/12599/files#r1668422877, we can move this part code out of mlflow/models/evaluation/default_evaluator.py , to make the code easier to maintain.

File: mlflow/models/evaluation/default_evaluator.py [hunk 4]
```diff
@@ -131,22 +130,15 @@ def _infer_model_type_by_labels(labels):
 
 def _extract_raw_model(model):
     model_loader_module = model.metadata.flavors["python_function"]["loader_module"]
+    # If we load a sklearn/xgboost model with mlflow.pyfunc.load_model, the model will be wrapped
+    # with a wrapper, we need to extract the raw model from it.
+    # Extract the raw model of xgboost flavor so that shap explainer uses the raw model
+    # instead of the wrapper and skip data schema validation.
```
Comment (B-Step62) [inline mlflow/models/evaluation/default_evaluator.py:136]:
```suggestion
    # If we load a model with mlflow.pyfunc.load_model, the model will be wrapped
    # with a pyfunc wrapper. We need to extract extract the raw model so that shap
    # explainer uses the raw model instead of the wrapper and skip data schema validation.
```

nit: minor corrections (L133: the wrapper is not only for sklean/xgboost, L135: we extract not only xgboost)

**Review by B-Step62 — APPROVED** [review]
LGTM!

File: mlflow/sklearn/__init__.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/sklearn/__init__.py:525]: I don't think we should modify the returned object here, which is unnatural side effect for the function named `get_raw_model`. I agree that the evaluator logic is super long and it can be structured better, but it should be in the evaluator namespace imo (or sth indicates it is used by evaluator) as it is only used for evaluation.

Comment (serena-ruan) [reply mlflow/sklearn/__init__.py:525]: I feel like we shouldn't modify the original model by adding a function on the raw xgboost model, while I moved the function into mlflow/xgboost/__init__.py file so we reuse it in the wrapper.

File: mlflow/models/evaluation/default_evaluator.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/models/evaluation/default_evaluator.py]: Updated according to above comment
